File: 2017/01_inverse_captcha/aoc01_ex/lib/aoc01.ex
defmodule AOC2017.Day01 do
  @moduledoc """
  Inverse Captcha

  Advent of Code 2017, day 1
  Solution by Geir Arne Hjelle, 2021-11-05
  """

  def main(args) do
    Enum.map(args, &solve/1)
  end

  def solve(path) do
    IO.puts("\nPart 1 (#{path}):")
    path |> parse() |> Enum.map(&part1/1) |> Enum.join("\n") |> IO.puts()

    IO.puts("\nPart 2: (#{path})")
    path |> parse() |> Enum.map(&part2/1) |> Enum.join("\n") |> IO.puts()
  end

  def parse(path) do
    with {:ok, file} <- File.read(path) do
      file
      |> String.trim()
      |> String.split("\n")
      |> Enum.map(&String.to_integer/1)
      |> Enum.map(&Integer.digits/1)
    end
  end

  def part1(input) do
    input
    |> prepend_last()
    |> Enum.reduce({nil, 0}, &compare_consecutive/2)
    |> elem(1)
  end

  def part2(input) do
    half_length = input |> length() |> half() |> trunc()

    input
    |> Enum.chunk_every(half_length)
    |> Enum.zip()
    |> Enum.filter(fn {first, second} -> first == second end)
    |> Enum.map(fn {first, second} -> first + second end)
    |> Enum.sum()
  end

  defp prepend_last(digits), do: [digits |> Enum.reverse() |> hd | digits]
  defp compare_consecutive(current, {current, total}), do: {current, total + current}
  defp compare_consecutive(current, {_previous, total}), do: {current, total}
  defp half(number), do: number / 2
end

File: old/2019/01_the_tyranny_of_the_rocket_equation/aoc01.jl
# The Tyranny of the Rocket Equation
#
# Advent of Code 2019, day 1
# Solution by Geir Arne Hjelle, 2020-11-30

module AOC01

using Pipe


# Calculate fuel based on mass
function fuel(mass)
    floor(mass / 3) - 2 |> Int64
end


# Calculate fuel based on mass, include fuel to cover mass of fuel
function recursive_fuel(mass)
    fuel_for_mass = mass |> fuel

    if fuel_for_mass <= 0
        return 0
    else
        return fuel_for_mass + recursive_fuel(fuel_for_mass)
    end
end


# Solve the problem for one file
function solve(input)
    # Parse input
    masses = @pipe split(input, "\n") .|> parse(Int64, _)

    # Part 1
    part_1 = masses .|> fuel |> sum

    # Part 2
    part_2 = masses .|> recursive_fuel |> sum

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module
